---
title: "Untitled"
author: "David Galbraith"
date: "10/28/2020"
output: html_document
---


```{r load_libraries}

library(Gviz)
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
library(data.table)
library(Cairo)
library(magrittr)
library(tidyverse)


source("./code/functions/build_locus_zoom.R")


##CONSTANTS
#treatments: Api, CAY, CUDC, LAQ, Pano, SB, SR42
treat <- "CAY"

#need to initialize these for some of the functions to work
x=1
y=1


```


```{r get_data}

#read the results file
df <- read.delim("./results/NCATS_ATAC_AllCombinedResultsAll.txt", header = T, sep = "\t")

#filter data for only significant genes and clean up the contrast column
df %<>%
  filter(Contrast == paste(treat, "_vs_NT", sep = ""),
         Significant == 1)

```


```{r get_annotation}
#get annotation
annot <- read.delim("./data/NCATS_annot.txt", header = T, sep = "\t")
```


```{r filter_annotation}

#identify genes of interest
#goi <- toupper(c("MYRFL", "PACS1", "CLPTM1L", "PRANCR"))
#top 5 genes
goi <- df %>%
  arrange(AdjP.ScoreInv, Distance) %>%
  distinct(Chr, Start, End, .keep_all = T) %>%
  head(n=5) %>%
  pull(Gene)

#filter annotation
annot %<>%
  filter(Gene %in% goi) %>%
  arrange(Gene, Start)

#increase the range of annotation using df to include peaks just outside of gene region
annot <- df %>%
  dplyr::select(Gene, Distance) %>%
  group_by(Gene) %>%
  mutate(maxUp = min(Distance),
         maxDown = max(Distance)) %>%
  ungroup() %>%
  left_join(annot, ., by = "Gene") %>%
  mutate(maxUp = ifelse(maxUp > 0, 0, maxUp),
         maxDown = ifelse(maxDown < 0, 0, maxDown),
         #expand the window by 500 in both directions to make the plotting look nicer
         Start = Start + maxUp - 500,
         End = End + maxDown + 500) %>%
  dplyr::select(-Distance, -maxUp, -maxDown) %>%
  distinct(.keep_all = T)
  


```


```{r Ideogram_track}
####### Ideogram track #######
itrack <- lapply(1:nrow(annot), function(x) 
  ideo_track(gen = 'hg38',
             chr = annot[x,1])
  )
```



```{r Annotation_track}
########## Annotation Track ############
########## Utilizes .bed file from Active Motif
########## plot specified medipseq peaks using an annotation track ##########

# extract MergedReg (medip-seq peaks) from 
# filtered bed file
all_peaks <- data.frame(df)

peaks <- lapply(1:nrow(annot), function(x)
  all_peaks %>%
    filter(as.character(Chr) == as.character(annot$Chr[x])) %>%
    #filter(Start >= annot$Start[x]) %>%
    #filter(End <= annot$End[x]) %>%
    filter(Gene %in% annot$Gene[x]) %>%
    mutate(Strand = annot$Strand[x])
  )

nam <- lapply(1:length(peaks), function(x) {
  y <- peaks[[x]] %>%
    pull(Gene)
  y <- unlist(sapply(1:length(y), function(z)
    paste(y[[z]], collapse = "\n")
    ))
  return(y)
  })

aTrack <- lapply(1:length(peaks), function(x) 
  aTrack <- anno_track(st = peaks[[x]]$Start,
                       chr = peaks[[x]]$Chr[1],
                       plot_name = nam[[x]],
                       #plot_name = "", # if you want to remove the names 
                       wid = abs(peaks[[x]]$Start - peaks[[x]]$End),
                       gen = "hg38",
                       title = "ATAC Peaks",
                       feat = "ATAC",
                       len = length(peaks[[x]]$Start))
)
```


#if you want the whole gene
```{r}

########## Genome Axis Track ############
range <- lapply(1:nrow(annot), function(x) {
  range <- annot[x,] %>%
    dplyr::slice(c(1, n())) %>%
    dplyr::select(Start, End)
  range <- c(range$Start[1], range$End[2])
})

gtrack <- lapply(1:length(range), function(x)
  GenomeAxisTrack(#\chr = peaks[[x]]$chr[1],
                  range=IRanges(start = range[[x]][1],
                                end = range[[x]][2])
                  )
  )

```



```{r Gene_Region_Track}
bmTrack <- lapply(1:nrow(annot), function(x)
  BiomartGeneRegionTrack(genome = "hg38",
                         chromosome = annot[x,"Chr"], 
                         start = annot[x,"Start"], 
                         end = annot[x,"End"],
                         #name = "ENSEMBL",
                         name = "", #to shrink the track
                         collapseTranscripts = "meta",
                         filters = list(ensembl_gene_id = annot[x,"ENSEMBL"]),
                         #using fill color from UCSC genome browser
                         fill = "#0C0C78",
                         col = "#0C0C78")
)
```



```{r Data_track}
########## Data Track ############
########## Utilizes .bam files from Active Motif or Rsubread
########## as input to calc coverage & plot ##########

# pull coverage from bam files
bam_files <- list.files("./data/bam/")
bam_files <- bam_files[grepl("i....bam", bam_files)]
bam_files <- bam_files[!grepl("\\.bai", bam_files)]
bam_files <- paste("./data/bam/", bam_files, sep = "")

bam_nam <- gsub(".*Astro-", "", bam_files)
bam_nam <- gsub(".*Astro_", "", bam_nam)
bam_nam <- gsub("_ATAC_.*", "", bam_nam)
bam_nam <- gsub("_", "-", bam_nam)


#reduce files
contrast <- c(treat, "NT")
contrast_files <- bam_files[c(grep(treat, bam_files), grep("NT", bam_files))]
contrast_nam <- bam_nam[c(grep(treat, bam_nam), grep("NT", bam_nam))]


# this has been reduced to only the target contrast, can do 1:6
all_cont <- lapply(1:length(range), function(x) {
  unlist(lapply(c(1:6), function(y) {
    build_cov(chr = peaks[[x]]$Chr[1],
              range1 = range[[x]][1],
              range2 = range[[x]][2],
              file = contrast_files[y])
  }), recursive = F)
})


```



```{r}
build_cov <- function(chr, range1, range2, file) {
  require(Rsamtools)
  # specify what chromosome and range of values 
  # to pull from bam (instead of pulling the entire
  # bam file, which is large & takes a long time)
  gr <- GRanges(seqnames = chr,
                ranges = IRanges(start = range1, end = range2))
  # specify a bam flag
  flag <- scanBamFlag(isUnmappedQuery = FALSE)
  # specify columns to pull from bam file (all of them)
  what <- c("qname", "flag", "rname", "strand", 
            "pos", "qwidth", "mapq", "cigar",
            "mrnm", "mpos", "isize", "seq", "qual")
  # set parameters for pulling  bam file: which chromo-
  # some and range, what columns from the bam file and
  # the flag
  param <- ScanBamParam(which = gr, what = what, flag = flag)
  # pull data from the bam file using parameters specified;
  # requires [[1]]
  bam <- scanBam(file,
                 param = param)[[1]]
  # create a IRanges object storing bam start and width data
  ranges <- IRanges(start=bam$pos, width=bam$qwidth, names=make.names(bam$qname, unique=TRUE))
  # create a GRanges object storing bam genomic locations and annotations
  ranges <- GRanges(seqnames=Rle(bam$rname), 
                    ranges=ranges, 
                    strand=Rle(bam$strand),
                    flag=bam$flag, 
                    readid=bam$rname )
  
  # split GRanges object by strand type (+ or -)
  # grs <- split(ranges, strand(ranges))
  # calc coverage for each ranges and associated
  # range 1 and 2
  # cov <- lapply(grs[c("+", "-")], function(y)
  #   coverage(ranges(y),
  #            width=end(gr)
  #   ))
  cov <- coverage(ranges(ranges), width = end(gr))
  
  # sort start/end coverages for positive strand
  # pos_st <- sort(unique(unlist(lapply(cov, function(y) c(start(y))))))
  # pos_end <- sort(unique(unlist(lapply(cov, function(y) c(end(y))))))
  
  pos_st <- sort(unique(c(start(cov))))
  pos_end <- sort(unique(c(end(cov))))
  
  # # if there are no coverage lengths on the positive strand
  # if(length(pos) == 0){
  #   
  #   mcols(gr) <- DataFrame(plus = 0, minus = 0)
  #   return(gr)
  # } else{
  
  # create an updated GRanges object with selected
  # chromosome, range1 & range2 limits w/two metadata
  # columns specifying coverage for plus/minus strands
  comb <- GRanges(seqnames = seqnames(gr)[1],
                  ranges = IRanges(start = pos_st[-1], end = pos_end[-1]),
                  coverage = as.numeric(cov[pos_st[-1]]))
  # }
  return(comb)
}
```


```{r Combine_Tracks}
all_dtrack_cont <- lapply(1:length(all_cont), function(x) {
  #get max coverage for consistent scales
  max_cov <- c()
  for(i in 1:length(all_cont[[x]])){
  tmp_max <- max(all_cont[[x]][[i]]$coverage)
  max_cov <- append(max_cov, tmp_max)
  }
  max_cov <- max(max_cov)
  ylims <- range(0, max_cov)
  #create data track
  unlist(lapply(c(1:6), function(y) 
    dat_track(all_cont[[x]][[y]],
              gen = "hg38",
              style = "histogram",
              title = contrast_nam[y],
              chr = gsub("chr", "", peaks[[x]]$Chr[1]),
              ylim = ylims
    )
    ), recursive = F)
})
```



```{r Plot_Tracks}

########## plot all data! ############
########## combine itrack, atrack and data tracks
########## into a single png
first <- lapply(1:length(itrack), function(x)
  list(itrack[[x]], aTrack[[x]], bmTrack[[x]]))
comb_cont <- lapply(1:length(all_dtrack_cont), function(x)
  c(first[[x]], all_dtrack_cont[[x]], gtrack[[x]]))

title_cont <- lapply(1:length(itrack), function(x)
  paste(treat, " vs NT at ", annot$Gene[x], " Gene", sep = ""))

plot_nam_cont <- lapply(1:nrow(annot), function(x)
  paste("./figures/gviz/Gene/Gene_", treat, "_vs_NT_", as.character(annot$Gene[x]), ".pdf", sep = "")
  )

lapply(1:length(comb_cont), function(x) {
  CairoPDF(file = plot_nam_cont[[x]], width = 10, height = 15, family = "NotoSans-Condensed")
  #reordering the tracks so the genome access track is near the top
  plotTracks(comb_cont[[x]][c(1, 10, 2:9)],
             main = title_cont[[x]],
             # set fill for y-axis (title)
             background.title = "white",
             # set font color for y-axis (title)
             col.title = "black",
             # set font family for y-axis (title)
             fontfamily.title = "NotoSans-Condensed",
             # set font size for y-axis (title)
             fontsize = 16,
             #sizes=c(0.5, 0.5, 0.5, 0.7, 0.75, 0.75, 0.75, 0.75, 0.75, 0.75),
             sizes=c(0.5, 0.5, 0.5, 0.3, 0.75, 0.75, 0.75, 0.75, 0.75, 0.75),
             from = annot[x,"Start"], 
             to = annot[x,"End"] 
  )
  graphics.off()
})

```

