---
title: "DESeq 2 analysis"
output: html_document
---

```{r load_libraries}
#load libraries
library(magrittr)
library(DESeq2)
library(readxl)
library(limma)
library(edgeR)
library(tidyverse)

#set a constant to attach to file names
outfileBase <- "NCATS_ATAC"

#set a common theme for plotting
mytheme <- theme(plot.title = element_text(lineheight = 0.8, face = "bold", size = 20),
                 axis.text = element_text(size = 14),
                 axis.title = element_text(face = "bold", colour = "Black", size = 16),
                 legend.text = element_text(colour = "Black", size = 12),
                 legend.title = element_text(colour = "Black", size = 14))
```



```{r get_data}
#read in the filtered data containing the normalized counts from Active Motif
df <- read.delim("./data/AM NCATS ATAC-Seq 8x3-samples 36926/AM NCATS ATAC-Seq 8x3-samples 36926/010ONIH_ATAC-Astro_mergedregs_NormCount_filtered.txt", header = T, row.names = 1, sep = "\t")
```

```{r}
#load merged region data to annotate reads 
mRegs <- read_xlsx("./data/AM NCATS ATAC-Seq diff analysis 36926/AM NCATS ATAC-Seq diff analysis 36926/010ONIH_ATAC-Astro_mergedregs_DESeq2.xlsx")

#removing space in the first column
colnames(mRegs)[1] <- "MergedRegion"

#select only the merged region, chromosome, and start and end positions
#select only those merged regions are in the 
mRegs %<>% 
  select(MergedRegion, Chr, Start, End) %>%
  filter(MergedRegion %in% rownames(df)) %>%
  mutate(MergedRegion = as.character(MergedRegion))
```


```{r create_design}
#create a design to tell DESeq2 which samples 
#belong to which treatments
doe <- data.frame("Sample" = colnames(df))
doe$Treatment <- gsub("\\..*", "", doe$Sample)
rownames(doe) <- doe$Sample
doe %<>% select(Treatment)
doe$Treatment <- as.factor(doe$Treatment)
doe$Treatment <- relevel(doe$Treatment, ref = "NT")
```


```{r log2_conversion}
#data is skewed, should take log2 to normalize
#this is normal for this type of data
hist(rowMeans(df))

#looks better
hist(rowMeans(log2(df)))
```


```{r}

#set up the DESeq object
dds <- DESeqDataSetFromMatrix(countData = df,
                              colData = doe,
                              design= ~ Treatment)

#set a filter for the minimum number of reads per sample
#if we filter for minimum of 10 reads/sample, total is 73758
#if we filter for minimum of 20 reads/sample, total is 33364
#We should probably use 10
keep <- rowSums(counts(dds))/ncol(dds) >= 10
dds <- dds[keep,]

#setting sizeFactors to 1 because this has already been done by Active Motif
#in their normalized count data file
sizeFactors(dds) <- c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)

#run DESeq
dds <- DESeq(object = dds)

#get contrasts to generate log fold change
#the -1 is to remove the intercept term
contrasts <- resultsNames(dds)[-1]

#initialize a list to add data
mylist <- list()

#loop through all of the contrasts
for (i in contrasts) {
  #extract results table from the DESeq analysis
  res <- results(dds, name = i, alpha = 0.05)
  #add shrunken log2 fold change and SE to the results table
  resShrink <- lfcShrink(dds, coef = i, type = "apeglm", res = res)
  
  #add the data to the list object created earlier
  mylist[[i]] <- as.data.frame(resShrink)
  
  #add merged region annotation to the results files and create
  #a consistent format to combine results
  resShrink %<>%
    as.data.frame() %>%
    rownames_to_column("MergedRegion") %>%
    mutate(MergedRegion = as.character(MergedRegion)) %>%
    left_join(mRegs, ., by = "MergedRegion") %>%
    filter_at(.vars = 9, all_vars(. < 0.05)) %>%
    select(Chr, Start, End, starts_with("baseMean"), starts_with("log2"), starts_with("lfc"), starts_with("pvalue"), starts_with("padj"))
  
  #write the results from each contrast to their own individual files
  write.table(resShrink, file = paste("./results/DESeq/", outfileBase, i, "_DESeqMACS.txt", sep = ""), row.names = F, sep = "\t", quote = F)
}

#combine information from all contrasts into one large data frame
ResAllContrasts <- do.call("cbind",mylist)

#the baseMean column from each treatment should be identical because it is
#calculated from all samples. This converts the name of the first one to
#"BaseMean" and removes the rest, then annotate with the merged region info
ResAllContrasts %<>%
  rename(BaseMean = Treatment_Api_vs_NT.baseMean) %>%
  select(BaseMean, contains("log2FoldChange"), contains("lfcSE"), contains("pvalue"), contains("padj")) %>%
  rownames_to_column("MergedRegion") %>%
  right_join(mRegs, ., by = "MergedRegion")

#write the data to a file
write.table(ResAllContrasts, "./results/DESeqMACS2filtered.txt", row.names = T, sep = "\t")
```




